golint package doesn't build

### Describe the bug
The `golint` package from `nixos-unstable` doesn't build

### Steps To Reproduce
Steps to reproduce the behavior:
1. `nix-env -iA nixos.golint` for instance
2. 
```console
% nix-env -iA nixos.golint
installing 'lint-20201208-83fdc39'
this derivation will be built:
  /nix/store/dsz5l71bnidizizf1isqrs8xcnz4fcwv-lint-20201208-83fdc39.drv
building '/nix/store/dsz5l71bnidizizf1isqrs8xcnz4fcwv-lint-20201208-83fdc39.drv'...
unpacking sources
unpacking source archive /nix/store/3w2lkbbqj11364g84y72pgy6svcp7dzg-lint-83fdc39
source root is lint-83fdc39
patching sources
configuring
building
Building subPackage .
golang.org/x/tools/go/ast/astutil
golang.org/x/tools/go/internal/gcimporter
golang.org/x/tools/go/gcexportdata
golang.org/x/lint
Building subPackage ./golint
golang.org/x/lint/golint
running tests
=== RUN   TestAll
--- PASS: TestAll (0.14s)
=== RUN   TestLine
--- PASS: TestLine (0.00s)
=== RUN   TestLintName
--- PASS: TestLintName (0.00s)
=== RUN   TestExportedType
--- PASS: TestExportedType (0.00s)
=== RUN   TestIsGenerated
--- PASS: TestIsGenerated (0.00s)
PASS
ok      golang.org/x/lint       0.149s
found packages pkg (4.go) and foo (blank-import-lib.go) in /build/lint-83fdc39/testdata
testdata/errorf-custom.go:9:2: cannot find package "." in:
        /build/lint-83fdc39/vendor/github.com/pkg/errors
error: builder for '/nix/store/dsz5l71bnidizizf1isqrs8xcnz4fcwv-lint-20201208-83fdc39.drv' failed with exit code 1;
       last 10 log lines:
       > --- PASS: TestLintName (0.00s)
       > === RUN   TestExportedType
       > --- PASS: TestExportedType (0.00s)
       > === RUN   TestIsGenerated
       > --- PASS: TestIsGenerated (0.00s)
       > PASS
       > ok   golang.org/x/lint       0.149s
       > found packages pkg (4.go) and foo (blank-import-lib.go) in /build/lint-83fdc39/testdata
       > testdata/errorf-custom.go:9:2: cannot find package "." in:
       >    /build/lint-83fdc39/vendor/github.com/pkg/errors
       For full logs, run 'nix log /nix/store/dsz5l71bnidizizf1isqrs8xcnz4fcwv-lint-20201208-83fdc39.drv'.
```

### Expected behavior
package builds

### Additional context
Happens also when installed as a dependency of some other packages via Home Manager.

### Notify maintainers
@jhillyerd 
@tomberek

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.17.1, NixOS, 22.05 (Quokka), 22.05pre367572.b6966d911da`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.7.0`
 - channels(miky): `"home-manager, nixos"`
 - channels(root): `"nixos, nixos-hardware"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

--- a/pkgs/development/go-modules/generic/default.nix
+++ b/pkgs/development/go-modules/generic/default.nix
@@ -204,7 +204,7 @@ let
         if [ -n "$subPackages" ]; then
           echo "$subPackages" | sed "s,\(^\| \),\1./,g"
         else
-          find . -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort --unique
+          find . -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort --unique | grep -v "$exclude"
         fi
       }
 
@@ -221,7 +221,6 @@ let
           export NIX_BUILD_CORES=1
       fi
       for pkg in $(getGoDirs ""); do
-        grep -q "$exclude" <<<$pkg && continue
         echo "Building subPackage $pkg"
         buildGoDir install "$pkg"
       done

--- a/pkgs/development/go-packages/generic/default.nix
+++ b/pkgs/development/go-packages/generic/default.nix
@@ -184,7 +184,7 @@ let
           echo "$subPackages" | sed "s,\(^\| \),\1$goPackagePath/,g"
         else
           pushd "$NIX_BUILD_TOP/go/src" >/dev/null
-          find "$goPackagePath" -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort | uniq
+          find "$goPackagePath" -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort | uniq | grep -v "$exclude"
           popd >/dev/null
         fi
       }
@@ -202,7 +202,6 @@ let
           export NIX_BUILD_CORES=1
       fi
       for pkg in $(getGoDirs ""); do
-        grep -q "$exclude" <<<$pkg && continue
         echo "Building subPackage $pkg"
         buildGoDir install "$pkg"
       done

